File: bench/py/65_struct_field_sum.py
# struct field-sum foreach: iterate a flat array<Struct> and sum its scalar
# fields, amplified by an outer rep loop. Exercises the direct struct-field read
# from the array bytes - distinct from 58_structs (mixed) and 64_struct_create
# (construction). __slots__ is the fairest CPython comparison to a POD struct.
import sys

scale = int(sys.argv[1]) if len(sys.argv) > 1 else 1


class Point3:
    __slots__ = ("x", "y", "z")

    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z


a = [Point3(i, i * 2, i * 3) for i in range(2000)]

s = 0
reps = 2000 * scale
for r in range(reps):
    for p in a:
        s = (s + p.x + p.y + p.z) % 1000000007

print("result:", s)

File: bench/cpp/65_struct_field_sum.cpp
/* SPDX-License-Identifier: BSD-2-Clause */
/* Faithful C++ of bench/my/65_struct_field_sum.my: iterate a flat array<Struct>
 * (2000 Point3) and sum its scalar fields, amplified by an outer rep loop.
 * std::vector<Point3> mirrors the flat POD-struct array; the field reads match
 * MyLang's direct-from-bytes reads. s carries a `%` per element (a serial
 * dependency, no close-form); bench_sink_ptr(a.data()) at the top of each pass
 * blocks hoisting the loop-invariant reduction out of the outer loop (cf. 18). */
#include "bench.h"

struct Point3 {
    long x;
    long y;
    long z;
};

int main(int argc, char **argv)
{
    long scale = bench_scale(argc, argv);

    std::vector<Point3> a;
    for (long i = 0; i < 2000; i++)
        a.push_back(Point3{i, i * 2, i * 3});

    long s = 0;
    long reps = 2000L * scale;
    for (long r = 0; r < reps; r++) {
        bench_sink_ptr(a.data());
        for (const Point3 &p : a)
            s = (s + p.x + p.y + p.z) % 1000000007;
    }

    printf("result: %ld\n", s);
    return 0;
}
